crates/matrix-sdk/src/widget_api/handler/capabilities.rs

```diff
@@ -0,0 +1,69 @@
+use async_trait::async_trait;
+use tokio::sync::mpsc::Receiver;
+
+use crate::widget_api::{
+    messages::{
+        capabilities::{EventFilter, Options},
+        from_widget::{
+            ReadEventRequest, ReadEventResponse, SendEventRequest, SendEventResponse,
+            SendToDeviceRequest,
+        },
+        MatrixEvent,
+    },
+    Result,
+};
+
+/// A wrapper for the matrix client that only exposes what is available through the capabilities.
+#[allow(missing_debug_implementations)]
+pub struct Capabilities {
+    // The options are a private member,
+    // they contain the type filters for the listeners.
+    filter_send_room_event: Vec<EventFilter>,
+    filter_read_room_event: Vec<EventFilter>,
+    filter_send_state_event: Vec<EventFilter>,
+    filter_read_state_event: Vec<EventFilter>,
```

Comment:
I'm still a bit skeptical about having the private members here that duplicate the state as part of the `event_reader` and `event_sender`. Why do we want to store it here? The filters are parts of the relevant event readers and senders, so instead of additionally copying it from there, I think we should access it from the `event_reader` and `event_sender` correspondingly. [Otherwise, we have that issue with the data types that I mentioned](https://github.com/toger5/matrix-rust-sdk/pull/14#discussion_r1284298870) that instead of having `f: X -> Y` (where Y is computable from X), we store both X and Y meaning that someone must keep them in sync, meaning that it introduces a possibility for bugs.

In fact, I was curious why you think storing them here would be easier, so I checked another PR to see how you implemented the `initialise() -> Capabilities` function to see how you handle the filter. Now check this out: [here](https://github.com/toger5/matrix-rust-sdk/pull/16/files#diff-febc64d55a9708f4d2cd63ac427fdd5ef55ae5b3dd9c45c4c59d2b9de1537ba6R36) we observe precisely the issue that I was talking about 🙂 (that if X and Y are stored both in the same structure despite the fact that Y is computable from X, there is a chance that data goes not synchronized and that someone must ensure that they match), and we have precisely that issue there:
1. Capabilities are created.
2. The `event_reader` and `event_sender` fields are set to the corresponding reader and sender.
3. However you forgot to set `filter_send_room_event` and the other 3 fields to the right value.
4. As a result we have a proper `event_reader` and `event_writer` (each of which already contains a filter), but the `filter_*` fields were not set, so constructing an `Options` object out of `Capabilities` will produce and invalid object that does not reflect the negotiated capabilities.

I suggest that we hedge against such risks by not storing these values in the structure as private members (avoiding all the risks and complexity that are connected with it) and simply extract them from **actual** handlers once they are negotiated, basically, the way I originally implemented it in on of the earlier commits [here](https://github.com/toger5/matrix-rust-sdk/pull/13/files#diff-a06241e78134bf5aa1a914e55b4f856e1706619fbce9598a0651ccf953b37f1bR63). What do you think?

Reply:
I think my though was, that we not only need it for the reader but also for the listener. But I think this is still possible/nicer if we put them seperate. let me try

Reply:
Oh wait but this is inside the handler. So those are the options. They definitly need to contain the information about the filter.

Reply:
This is addressed in this follow up commit: https://github.com/toger5/matrix-rust-sdk/pull/16/commits/b3541bda783a69cdb1ebf455764fbca9785e89c6

Reply:
Awesome!